docs/contributing/inter-thread-context-propagation.md

```diff
@@ -0,0 +1,87 @@
+# The story of context propagation across threads
+
+## The need
+Take a look at the following two pseudo-code snippets (see below for explanations).
+
+```
+Executor pool = Executors.newFixedThreadPool(10)
+
+public void doGet(HttpServletRequest request, HttpServletResponse response) {
+    Future f1 = pool.submit(() -> {
+        return userRepository.queryShippingAddress(requet)
+    })
+    Future f2 = pool.submit(() -> {
+        return warehouse.currentState(requet)
+    })
+    writeResponse(response, f1.get(), f2.get())
+}
+```
+
+```
+Executor pool = Executors.newFixedThreadPool(10)
+
+public void doGet(HttpServletRequest request, HttpServletResponse response) {
+    final AsyncContext acontext = request.startAsync();
+    acontext.start(() -> {
+            String address = userRepository.queryShippingAddress(requet)
+            HttpServletResponse response = acontext.getResponse();
+            writeResponse(response, address)
+            acontext.complete();
+   }
+}
+```
+
+In both cases request processing requires some potentially long operation and application developer
+wants to do them off the main thread. In the first case this hand-off between request accepting thread
+and request processing thread happens manually, by submitting work into some thread pool.
+In the second case it is the framework that handles separate thread pool and passing work to it.
+
+In cases like this proper tracing solution should still combine into a single trace all the work
+required for request processing, irregardless in what thread that work happened. With proper
+parent-child relationship between span: span representing shipping address query should be the child
+of the span which denotes accepting HTTP request.
+
+## The solution
+Java auto instrumentation uses an obvious solution to the requirement above: we attach current execution
+context (represented in the code by `io.grpc.Context`) with each `Runnable`, `Callable` and `ForkJoinTask`.
+"Current" means the context active on the thread which calls `Executor.execute` (and its analogues
+such as `submit`, `invokeAll` etc) at the moment of that call. Whenever some other thread starts
+actual execution of that `Runnable` (or `Callable` or `ForkJoinTask`), that context get restored
+on that thread for the duration of the execution. This can be illustrated by the following pseudo-code:
+
+```
+    var job = () -> {
+        try(Scope scope = withScopedContext(this.context)) {
+            return userRepository.queryShippingAddress(requet)
+        }}
+    job.context = Context.current()
+    Future f1 = pool.submit()
+
+```
+
+## The drawback
+There are some runtime environments which, simplifying, do the following:
```

Comment:
the example in this section shows how if we leak to the first `pool.submit`, then it will keep leaking, but i don't think it answers the question of how we leak to the first `pool.submit`.

maybe the more concrete example you tracked down with tomcat / async dispatch would be more illustrative of the problem?

Reply:
I don't quite understand what do you mean by "leak the first pool.submit"? My idea was to show that any context, which was active during the first `pool.submit` will continue to be active for all runnables submitted via this started chain. That second `AcceptRequestRunnable` will have the same active context as the first one.

Reply:
I'm not clear, are you showing that the context (e.g. `spanId`) from `REQUEST#1` below will leak to `REQUEST#2`? I don't think it will, at least I don't think it will via the `java-concurrent` instrumentation alone.

This is what it looks like to me:

```
pool.submit(new AcceptRequestRunnable() {    <!--- assuming clean context to start with
    Request req = readRequest()    <--- REQUEST#1
    pool.submit(new ProcessRequestRunnable(req) {    <!--- the clean context from above will be propagated
        writeResponse(process(req))    <--- REQUEST#1
        pool.submit(new AcceptRequestRunnable() {    <!--- the clean context from above will be propagated
            Request req = readRequest()    <--- REQUEST#2
            ... repeat until shutdown
    })
})
```

Reply:
I tried to add clarifications, PTAL